File: src/sim/pv/dc.ts
import type { Celsius, WattsPerM2 } from '../../types/units'

/** PVWatts v5 default temperature coefficient of power for crystalline silicon, per degree C */
export const PVWATTS_GAMMA_PDC_PER_C = -0.0047

export const PVWATTS_REFERENCE_TEMP_C = 25 as Celsius

const STC_IRRADIANCE_W_M2 = 1000

/**
 * PVWatts v5 DC model: Pdc = (G_poa / 1000) * Pdc0 * (1 + gamma * (Tcell - 25)).
 * Pdc0 is the array DC nameplate, so the result is in the same unit
 */
export const pvwattsDc = (
  poaWM2: WattsPerM2,
  cellTempC: Celsius,
  nameplateDc: number,
  gammaPdcPerC: number,
): number =>
  poaWM2 <= 0
    ? 0
    : Math.max(
        0,
        (poaWM2 / STC_IRRADIANCE_W_M2) *
          nameplateDc *
          (1 + gammaPdcPerC * (cellTempC - PVWATTS_REFERENCE_TEMP_C)),
      )

File: crates/agv-sim/src/pv/dc.rs
//! The PVWatts v5 DC model, ported from `src/sim/pv/dc.ts`.

/// PVWatts v5 temperature coefficient of power for crystalline silicon, per degree C.
pub const PVWATTS_GAMMA_PDC_PER_C: f64 = -0.0047;
pub const PVWATTS_REFERENCE_TEMP_C: f64 = 25.0;
const STC_IRRADIANCE_W_M2: f64 = 1000.0;

/// `Pdc = (G_poa / 1000) * Pdc0 * (1 + gamma * (Tcell - 25))`.
///
/// `nameplate_dc` is the array DC nameplate, so the result carries whatever unit it was given.
pub fn pvwatts_dc(poa_wm2: f64, cell_temp_c: f64, nameplate_dc: f64, gamma_pdc_per_c: f64) -> f64 {
    if poa_wm2 <= 0.0 {
        return 0.0;
    }
    ((poa_wm2 / STC_IRRADIANCE_W_M2)
        * nameplate_dc
        * (1.0 + gamma_pdc_per_c * (cell_temp_c - PVWATTS_REFERENCE_TEMP_C)))
        .max(0.0)
}
